# Acquire the Data

In [1]:
import numpy as np
import pandas as pd
import requests, zipfile, io

## Get the Files

In [2]:
# Get the url for ml-100k
zip_file_url = "http://files.grouplens.org/datasets/movielens/ml-100k.zip"

In [3]:
# Get the files and extract them
print("Downloading movielens data...")
r = requests.get(zip_file_url, stream=True)
z = zipfile.ZipFile(io.BytesIO(r.content))
z.extractall()
print("Done.")

Done.


## Users, Items & Ratings

**Users** from the README


`
u.user     -- Demographic information about the users; this is a tab
              separated list of
              user id | age | gender | occupation | zip code
              The user ids are the ones used in the u.data data set.
`


In [20]:
users_cols = ['user_id', 'age', 'sex', 'occupation', 'zip_code']
users = pd.read_csv('ml-100k/u.user', sep='|', names=users_cols, encoding='latin-1')

In [21]:
users.shape

(943, 5)

In [22]:
users.head()

user_id  age sex  occupation zip_code
0        1   24   M  technician    85711
1        2   53   F       other    94043
2        3   23   M      writer    32067
3        4   24   M  technician    43537
4        5   33   F       other    15213

**Items** from the README

`
u.item     -- Information about the items (movies); this is a tab separated
              list of
              movie id | movie title | release date | video release date |
              IMDb URL | unknown | Action | Adventure | Animation |
              Children's | Comedy | Crime | Documentary | Drama | Fantasy |
              Film-Noir | Horror | Musical | Mystery | Romance | Sci-Fi |
              Thriller | War | Western |
              The last 19 fields are the genres, a 1 indicates the movie
              is of that genre, a 0 indicates it is not; movies can be in
              several genres at once.
              The movie ids are the ones used in the u.data data set.
`

In [23]:
genre_cols = [
    "genre_unknown", "Action", "Adventure", "Animation", "Children", "Comedy",
    "Crime", "Documentary", "Drama", "Fantasy", "FilmNoir", "Horror",
    "Musical", "Mystery", "Romance", "SciFi", "Thriller", "War", "Western"
]
items_cols = ['movie_id', 'title', 'release_date', "video_release_date", "imdb_url"] + genre_cols
items_raw = pd.read_csv('ml-100k/u.item', sep='|', names=items_cols, encoding='latin-1')

In [24]:
items_raw.shape

(1682, 24)

In [25]:
items_raw.head()

movie_id              title release_date  video_release_date  \
0         1   Toy Story (1995)  01-Jan-1995                 NaN   
1         2   GoldenEye (1995)  01-Jan-1995                 NaN   
2         3  Four Rooms (1995)  01-Jan-1995                 NaN   
3         4  Get Shorty (1995)  01-Jan-1995                 NaN   
4         5     Copycat (1995)  01-Jan-1995                 NaN   

                                            imdb_url  genre_unknown  Action  \
0  http://us.imdb.com/M/title-exact?Toy%20Story%2...              0       0   
1  http://us.imdb.com/M/title-exact?GoldenEye%20(...              0       1   
2  http://us.imdb.com/M/title-exact?Four%20Rooms%...              0       0   
3  http://us.imdb.com/M/title-exact?Get%20Shorty%...              0       1   
4  http://us.imdb.com/M/title-exact?Copycat%20(1995)              0       0   

   Adventure  Animation  Children  ...  Fantasy  FilmNoir  Horror  Musical  \
0          0          1         1  ...        0         0       0        0   
1          1          0         0  ...        0         0       0        0   
2          0          0         0  ...        0         0       0        0   
3          0          0         0  ...        0         0       0        0   
4          0          0         0  ...        0         0       0        0   

   Mystery  Romance  SciFi  Thriller  War  Western  
0        0        0      0         0    0        0  
1        0        0      0         1    0        0  
2        0        0      0         1    0        0  
3        0        0      0         0    0        0  
4        0        0      0         1    0        0  

[5 rows x 24 columns]

**Ratings** from the README

`
u.data     -- The full u data set, 100000 ratings by 943 users on 1682 items.
              Each user has rated at least 20 movies.  Users and items are
              numbered consecutively from 1.  The data is randomly
              ordered. This is a tab separated list of 
              user id | item id | rating | timestamp. 
              The time stamps are unix seconds since 1/1/1970 UTC 
`

In [26]:
ratings_cols = ['user_id', 'movie_id', 'rating', 'unix_timestamp']
ratings = pd.read_csv('ml-100k/u.data', sep='\t', names=ratings_cols, encoding='latin-1')

In [27]:
ratings.shape

(100000, 4)

In [28]:
ratings.head()

user_id  movie_id  rating  unix_timestamp
0      196       242       3       881250949
1      186       302       3       891717742
2       22       377       1       878887116
3      244        51       2       880606923
4      166       346       1       886397596

## Save the Data

In [29]:
from utils import create_directory

In [30]:
### Create Feature Directory
create_directory("/data")

Directory already exists /Users/amitkaps/Documents/github/recommendation/MovieLens/data


In [31]:
# Save the data
users.to_csv("data/users.csv", index=None)
items_raw.to_csv("data/items_raw.csv", index=None)
ratings.to_csv("data/ratings.csv", index=False)